src/federation.rs

```diff
@@ -0,0 +1,268 @@
+                            threshold: (self.threshold - 1) as usize,
+                            share_count: self.nodevss.len(),
+                        },
+                        commitments: vss
+                            .positive_commitments
+                            .iter()
+                            .map(|i| i.to_point())
+                            .collect(),
+                    },
+                    secret_share: vss.positive_secret,
+                },
+            );
+        }
+        secret_shares
+    }
+
+    /// Returns an aggregated share of the node.
+    pub fn node_secret_share(&self) -> SharedKeys {
+        let secret_shares = self.node_shared_secrets();
+
+        let shared_keys =
+            Sign::verify_vss_and_construct_key(&secret_shares, &(self.node_index() + 1))
+                .expect("invalid vss");
+        shared_keys
+    }
+
+    pub fn validate(&self) -> Result<(), Error> {
+        // Check all sender is different.
+        let signers = self.signers();
+        let is_overlap = signers
+            .iter()
+            .any(|i| signers.iter().filter(|j| i == *j).count() > 1);
+        if is_overlap {
+            return Err(Error::InvalidFederation(
+                "nodevss has overlapping sender vss.".to_string(),
+            ));
+        }
+
+        // Check all receiver is the node itself.
+        if self
+            .nodevss
+            .iter()
+            .any(|i| i.receiver_public_key != self.signer_id.pubkey)
+        {
+            return Err(Error::InvalidFederation("The nodevss has wrong receiver value. All VSS's receiver_public_key should be equal with publich key of the signer who runs the node.".to_string()));
+        }
+
+        // Check all commitment length is correct.
+        if self
+            .nodevss
+            .iter()
+            .any(|vss| vss.positive_commitments.len() != self.threshold as usize)
+        {
+            return Err(Error::InvalidFederation(
+                "The nodevss has wrong vss which has wrong number of commitments.".to_string(),
+            ));
+        }
+
+        // verify each vss.
+        if Sign::verify_vss_and_construct_key(&self.node_shared_secrets(), &(self.node_index() + 1))
+            .is_err()
+        {
+            return Err(Error::InvalidFederation(
+                "The nodevss includes invalid share.".to_string(),
+            ));
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::errors::Error;
+    use crate::federation::{Federation, Federations};
+    use crate::net::SignerID;
+    use crate::tests::helper::keys::TEST_KEYS;
+    use crate::tests::helper::node_vss::node_vss;
+    use curv::elliptic::curves::traits::ECScalar;
+
+    #[test]
+    fn test_signers() {
+        let federation = Federation::new(TEST_KEYS.pubkeys()[4], 0, 3, node_vss(0));
+
+        let mut pubkeys = TEST_KEYS.pubkeys();
+        pubkeys.sort_by(|a, b| {
```

Comment:
Could you use NodeParameters::sort_publickey()?

Reply:
I want to keep this code.

Because I want to eliminate `NodeParameters::sort_publickey()` in sometime after merged this RP. After that, all the place where needs sorted pubkey should use `Federation::signers()`.
What do you think?

Reply:
No problem.